File: src/entropy/Container.cpp
#include <entropy/Container.h>

#include <iostream>
#include <assert.h>

# define MIN(a,b) (((a)<(b))?a:b)

using namespace std;

Container::Container(int rows, int columns)
{
  _data         = new double*[rows];
  _domains      = new double*[columns];
  _bins         = new int[columns];
  _mode         = CONTAINER_DISCRETISE_UNIFORM;

  _domainsGiven = false;
  _binsGiven    = false;

  for(int i = 0; i < columns; i++)
  {
    _domains[i] = new double[2];
  }

  for(int i = 0; i < rows; i++)
  {
    _data[i] = new double[columns];
  }

  for(int r = 0; r < rows; r++)
  {
    for(int c = 0; c < columns; c++)
    {
      _data[r][c] = 0.0;
    }
  }

  _rows      = rows;
  _columns   = columns;
  _fillIndex = 0;
}

Container::~Container()
{
  for(int r = 0; r < _rows;    r++) delete _data[r];
  for(int c = 0; c < _columns; c++) delete _domains[c];

  delete _data;
  delete _domains;
  delete _bins;
}

double Container::operator()(const int row, const int column) const
{
  assert(row    < _rows);
  assert(column < _columns);
  return _data[row][column];
}

const Container& Container::operator<<(const double &value) const
{
  assert(_fillIndex < _rows * _columns);

  int c = _fillIndex % _columns;
  int r = _fillIndex / _columns;

  _data[r][c] = value;
  Container *co = (Container*)this;
  co->_fillIndex++;

  return *this;
}

int Container::rows()
{
  return _rows;
}

int Container::columns()
{
  return _columns;
}

double Container::get(int row, int column)
{
  assert(row    < _rows);
  assert(column < _columns);
  return _data[row][column];
}

void Container::setBinSizes(int *bins)
{
  for(int c = 0; c < _columns; c++)
  {
    _bins[c] = bins[c];
  }
  _binsGiven = true;
}

void Container::setDomains(double **domains)
{
  for(int c = 0; c < _columns; c++)
  {
    _domains[c][0] = domains[c][0];
    _domains[c][1] = domains[c][1];
  }
  _domainsGiven = true;
}


Container* Container::discretise()
{
  switch(_mode)
  {
    case CONTAINER_DISCRETISE_UNIFORM:
      return __uniformDiscretisation();
      break;
    default:
      cerr << "Unknown discretisation mode given: " << _mode << endl;
      break;
  }
  return NULL;
}

void Container::setDiscretisationMode(int mode)
{
  _mode = mode;
}

int Container::__discretiseAndCombineValues(double *values)
{
  assert(_domainsGiven && _binsGiven);

  int value  = 0;
  int factor = 1;

  for(int c = 0; c < _columns; c++)
  {
    double mapped = ((values[c]      - _domains[c][0]) /
                     (_domains[c][1] - _domains[c][0]) * _bins[c]);
    double cropped = MIN(_bins[c]-1, mapped);
    if(c > 0) factor *= _bins[c-1];
    value += (int)(cropped * factor);
  }

  return value;
}

Container* Container::__uniformDiscretisation()
{
  Container *copy = new Container(_rows, 1);
  for(int r = 0; r < _rows; r++)
  {
    double v = (double)__discretiseAndCombineValues(_data[r]);
    (*copy) << v;
  }
  return copy;
}

File: py/entropy/container.py
import numpy

class Container:
    data          = None
    fillIndex     = 0
    addByRow      = True
    rows          = -1
    cols          = -1
    isDiscretised = False

    def __init__(self, rows, cols):
        self.fillIndex     = 0
        self.data          = numpy.zeros((rows, cols))
        self.rows          = rows
        self.cols          = cols
        self.addByRow      = True
        self.isDiscretised = False

    def setAddByRow(self):
        self.addByRow = True

    def setAddByColumn(self):
        self.addByRow = False
        
    def add(self, valum):
        c = -1
        r = -1
        if self.addByRow is True:
            c = self.fillIndex % self.cols
            r = self.fillIndex / self.cols
        else:
            c = self.fillIndex / self.rows
            r = self.fillIndex % self.rows

        self.data[r,c] = valum
        self.fillIndex = self.fillIndex + 1

    def shape(self):
        return self.data.shape

    def setBinsSizes(self, lst):
        self.bins = lst

    def setDomains(self, lst_of_tuples):
        self.domains = lst_of_tuples

    def discretise(self):
        self.discretiseByColumn()
        self.relabel() # count from 0,1,..., i.e. rename the bins
        self.combineDiscretisedColumns()
        self.relabel() # count from 0,1,..., i.e. rename the bins

    def discretiseByColumn(self):
        # currently only uniform binning is supported
        self.uniformDiscretisationByColumn()

    def uniformDiscretisationByColumn(self):
        for c in range(0, self.cols):
            for r in range(0, self.rows):
                assert (self.domains[c][0] <= self.data[r,c] and self.data[r,c] <= self.domains[c][1])
                mapped  = int(((self.data[r,c] - self.domains[c][0]) / (self.domains[c][1] - self.domains[c][0])) * self.bins[c])
                cropped = int(min(self.bins[c]-1, mapped))
                self.data[r,c] = cropped
        self.isDiscretised = True

    def relabel(self):
        for c in range(0, self.cols):
            values = []
            for r in range(0, self.rows):
                try:
                    index = values.index(self.data[r,c])
                except ValueError:
                    index = len(values)
                    values.append(self.data[r,c])
                self.data[r,c] = index


    def combineDiscretisedColumns(self):
        new_data = numpy.zeros((self.rows,1))
        for r in range(0, self.rows):
            v = 0
            f = 1
            for c in range(0, self.cols):
                if c > 0:
                    f  = f * self.bins[c-1]
                v += f * self.data[r,c]
            new_data[r,0] = v
        self.data = new_data
        self.cols = 1
